src/magic_folder/downloader.py

```diff
@@ -0,0 +1,578 @@
+                while q and not ancestor:
+                    snap = q.popleft()
+                    for parent_cap in snap.parents_raw:
+                        if existing_snap.capability == parent_cap:
+                            ancestor = True
+                            break
+                        else:
+                            q.append(self._remote_cache.cached_snapshots[parent_cap])
+                action.add_success_fields(ancestor=ancestor)
+
+                # whether we found a common ancestor or not
+                # tells us if we've got an overwrite or
+                # conflict
+                if ancestor:
+                    self._magic_fs.mark_overwrite(snapshot, staged)
+                else:
+                    self._magic_fs.mark_conflict(snapshot, staged)
+
+            else:
+                # there is no local file
+                # XXX we don't handle deletes yet
+                assert not local_snap and not remote_snap, "Internal inconsistency"
+                self._magic_fs.mark_overwrite(snapshot, staged)
+
+            # remember the last remote we've downloaded
+            self._config.store_remotesnapshot(snapshot.name, snapshot)
+
+            # XXX careful here, we still need something that makes
+            # sure mismatches between remotesnapshots in our db and
+            # the Personal DMD are reconciled .. that is, if we crash
+            # here and/or can't update our Personal DMD we need to
+            # retry later.
+            yield self.tahoe_client.add_entry_to_mutable_directory(
+                self._config.upload_dircap,
+                snapshot.name,
+                snapshot.capability.encode("ascii"),
+                replace=True,
+            )
+
+
+@implementer(IMagicFolderFilesystem)
+@attr.s
+class LocalMagicFolderFilesystem(object):
+    """
+    Makes changes to a local directory.
+    """
+
+    magic_path = attr.ib(validator=instance_of(FilePath))
+    staging_path = attr.ib(validator=instance_of(FilePath))
+
+    @inlineCallbacks
+    def download_content_to_staging(self, remote_snapshot, tahoe_client):
+        """
+        IMagicFolderFilesystem API
+        """
+        import hashlib
+        h = hashlib.sha256()
+        h.update(remote_snapshot.capability)
+        staged_path = self.staging_path.child(h.hexdigest())
+        with staged_path.open('wb') as f:
+            yield tahoe_client.stream_capability(remote_snapshot.content_cap, f)
+        returnValue(staged_path)
+
+    def mark_overwrite(self, remote_snapshot, staged_content):
+        """
+        This snapshot is an overwrite. Move it from the staging area over
+        top of the existing file (if any) in the magic-folder.
+
+        :param FilePath staged_content: a local path to the downloaded
+            content.
+        """
+        local_path = self.magic_path.preauthChild(magic2path(remote_snapshot.name))
+        staged_content.moveTo(local_path)
```

Comment:
`moveTo` may involve copying all the content synchronously. We may want to avoid causing the process to block if we end up copying a multi-GB files here. (We should at least file an issue about this).

See also the comment above about race-conditions when this isn't synchronous.